src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -23,24 +21,35 @@ public partial class Http2Stream : HttpProtocol
     {
         private readonly Http2StreamContext _context;
         private readonly Http2OutputProducer _http2Output;
-        private readonly Http2StreamOutputFlowControl _outputFlowControl;
-        private int _requestAborted;
+        private readonly StreamInputFlowControl _inputFlowControl;
+        private readonly StreamOutputFlowControl _outputFlowControl;
+
+        private StreamCompletionFlags _completionState;
+        private readonly object _completionLock = new object();
 
         public Http2Stream(Http2StreamContext context)
             : base(context)
         {
             _context = context;
-            _outputFlowControl = new Http2StreamOutputFlowControl(context.ConnectionOutputFlowControl, context.ClientPeerSettings.InitialWindowSize);
+
+            _inputFlowControl = new StreamInputFlowControl(
+                _context.StreamId,
+                _context.FrameWriter,
+                context.ConnectionInputFlowControl,
+                Http2PeerSettings.DefaultInitialWindowSize,
+                Http2PeerSettings.DefaultInitialWindowSize / 2);
+
+            _outputFlowControl = new StreamOutputFlowControl(context.ConnectionOutputFlowControl, context.ClientPeerSettings.InitialWindowSize);
             _http2Output = new Http2OutputProducer(context.StreamId, context.FrameWriter, _outputFlowControl, context.TimeoutControl, context.MemoryPool);
+
+            RequestBodyPipe = CreateRequestBodyPipe();
             Output = _http2Output;
         }
 
         public int StreamId => _context.StreamId;
 
         public bool RequestBodyStarted { get; private set; }
-        public bool EndStreamReceived { get; private set; }
-
-        protected IHttp2StreamLifetimeHandler StreamLifetimeHandler => _context.StreamLifetimeHandler;
+        public bool EndStreamReceived => (_completionState & StreamCompletionFlags.EndStreamReceived) == StreamCompletionFlags.EndStreamReceived;
```

Comment:
I see what you meant by wanting to use HasFlag

Reply:
What's wrong with HasFlag? Is an intrinsic now https://github.com/dotnet/coreclr/pull/13748

Reply:
The method exists in desktop though? Its just slower?

Reply:
Yep. The desktop version of HasFlag does reflection each time AFAIK.